# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFont.cpp
# Commit: 673f9010ea75
# Full Hash: 673f9010ea75382463067acf3da66683f4eecae1
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2022-08-17 09:10:29
# Regressor Bug: 1740530
# File Overlap Count: 1
# Description:
#   Bug 1740530 - patch 3 - Replace GetColorGlyphLayers and the layer-rendering code in gfxFont with a zero-copy implementation in COLRFonts. r=gfx-reviewers,lsalzman
#   
#   Depends on D152037
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D152038
# ==============================================================================

diff -r 429601610f26 -r 673f9010ea75 gfx/thebes/gfxFont.cpp
--- a/gfx/thebes/gfxFont.cpp	Tue Aug 16 12:48:32 2022 +0000
+++ b/gfx/thebes/gfxFont.cpp	Tue Aug 16 12:48:32 2022 +0000
@@ -2464,61 +2464,23 @@
 
 bool gfxFont::RenderColorGlyph(DrawTarget* aDrawTarget, gfxContext* aContext,
                                layout::TextDrawTarget* aTextDrawer,
-                               mozilla::gfx::ScaledFont* scaledFont,
-                               mozilla::gfx::DrawOptions aDrawOptions,
-                               const mozilla::gfx::Point& aPoint,
+                               ScaledFont* aScaledFont,
+                               DrawOptions aDrawOptions, const Point& aPoint,
                                uint32_t aGlyphId) const {
-  AutoTArray<uint16_t, 8> layerGlyphs;
-  AutoTArray<mozilla::gfx::DeviceColor, 8> layerColors;
-
-  mozilla::gfx::DeviceColor ctxColor;
-  if (!aContext->GetDeviceColor(ctxColor)) {
-    ctxColor = ToDeviceColor(mozilla::gfx::sRGBColor::OpaqueBlack());
-  }
-  if (!COLRFonts::GetColorGlyphLayers(GetFontEntry()->GetCOLR(),
-                                      GetFontEntry()->GetCPAL(), aGlyphId,
-                                      ctxColor, layerGlyphs, layerColors)) {
-    return false;
+  DeviceColor ctxColor;
+  sRGBColor currentColor = aContext->GetDeviceColor(ctxColor)
+                               ? sRGBColor::FromABGR(ctxColor.ToABGR())
+                               : sRGBColor::OpaqueBlack();
+
+  if (const auto* layers =
+          COLRFonts::GetGlyphLayers(GetFontEntry()->GetCOLR(), aGlyphId)) {
+    return COLRFonts::PaintGlyphLayers(GetFontEntry()->GetCOLR(),
+                                       GetFontEntry()->GetCPAL(), layers,
+                                       aDrawTarget, aTextDrawer, aScaledFont,
+                                       aDrawOptions, currentColor, aPoint);
   }
 
-  // Default to opaque rendering (non-webrender applies alpha with a layer)
-  float alpha = 1.0;
-  if (aTextDrawer) {
-    // defaultColor is the one that comes from CSS, so it has transparency info.
-    bool hasComplexTransparency = 0.f < ctxColor.a && ctxColor.a < 1.f;
-    if (hasComplexTransparency && layerGlyphs.Length() > 1) {
-      // WebRender doesn't support drawing multi-layer transparent color-glyphs,
-      // as it requires compositing all the layers before applying transparency.
-      // (pretend to succeed, output doesn't matter, we will emit a blob)
-      aTextDrawer->FoundUnsupportedFeature();
-      return true;
-    }
-
-    // If we get here, then either alpha is 0 or 1, or there's only one layer
-    // which shouldn't have composition issues. In all of these cases, applying
-    // transparency directly to the glyph should work perfectly fine.
-    //
-    // Note that we must still emit completely transparent emoji, because they
-    // might be wrapped in a shadow that uses the text run's glyphs.
-    alpha = ctxColor.a;
-  }
-
-  for (uint32_t layerIndex = 0; layerIndex < layerGlyphs.Length();
-       layerIndex++) {
-    Glyph glyph;
-    glyph.mIndex = layerGlyphs[layerIndex];
-    glyph.mPosition = aPoint;
-
-    mozilla::gfx::GlyphBuffer buffer;
-    buffer.mGlyphs = &glyph;
-    buffer.mNumGlyphs = 1;
-
-    mozilla::gfx::DeviceColor layerColor = layerColors[layerIndex];
-    layerColor.a *= alpha;
-    aDrawTarget->FillGlyphs(scaledFont, buffer, ColorPattern(layerColor),
-                            aDrawOptions);
-  }
-  return true;
+  return false;
 }
 
 bool gfxFont::HasColorGlyphFor(uint32_t aCh, uint32_t aNextCh) {
@@ -2544,7 +2506,7 @@
   }
   // Check if there is a COLR/CPAL or SVG glyph for this ID.
   if (fe->TryGetColorGlyphs() &&
-      COLRFonts::HasColorLayersForGlyph(fe->GetCOLR(), gid)) {
+      COLRFonts::GetGlyphLayers(fe->GetCOLR(), gid)) {
     return true;
   }
   if (fe->TryGetSVGData(this) && fe->HasSVGGlyph(gid)) {